### ISSUE TITLE
Poor performance.

### ISSUE THREAD

**Issue by hadialqattan** [hakancelik96/unimport#90]
https://github.com/hakancelik96/unimport/blob/c70fbcfe6c232b07a65b3ec8efa17e6039e60f33/unimport/scan.py#L192-L204

This function iterates over the whole comments, while there is a param `type_comments` that can be passed to `ast.parse` in order to get function/assign type comment efficiently!

If you agree with me please assign this issue to me, if not please tell me why or/and close this issue!

**Comment by hakancelik96** [comment]
`type_comments` only available python3.8 and higher

**Comment by hakancelik96** [comment]
Hi, Yes I agree with you and I did an experiment, all tests worked, I leave you diff, it can help you with the PR you will open.
Thank you for noticing and reporting this issue, I look forward to your PR, impatiently.

```diff
diff --git a/tests/test_refactor.py b/tests/test_refactor.py
index d624c6c..66f4694 100644
--- a/tests/test_refactor.py
+++ b/tests/test_refactor.py
@@ -655,6 +655,36 @@ class TestTyping(RefactorTestCase):
             self.session.refactor(action),
         )

+    @unittest.skipIf(
+        not PY38_PLUS, "This feature is only available for python 3.8."
+    )
+    def test_type_comments_with_variable(self):
+        action = (
+            "from typing import List\n"
+            "test_variable = [2] # type: List[int]\n"
+        )
+        self.assertEqual(
+            action,
+            self.session.refactor(action),
+        )
+
+    @unittest.skipIf(
+        not PY38_PLUS, "This feature is only available for python 3.8."
+    )
+    def test_type_comment_params(self):
+        action = (
+            "from typing import List\n"
+            "def x(\n"
+            "   f: # type:List,\n"
+            "   r: # type:str\n"
+            "):\n"
+            "   pass\n"
+        )
+        self.assertEqual(
+            action,
+            self.session.refactor(action),
+        )
+
     def test_variable(self):
         action = "from typing import List, Dict\n" "test: 'List[Dict]'\n"
         self.assertEqual(
diff --git a/unimport/scan.py b/unimport/scan.py
index 497b6a8..f3a9d9c 100644
--- a/unimport/scan.py
+++ b/unimport/scan.py
@@ -103,6 +103,8 @@ class Scanner(ast.NodeVisitor):
             and not self.include_star_import
         ):
             self.functions.append(Name(lineno=node.lineno, name=node.name))
+        if hasattr(node, "type_comment") and node.type_comment:
+            self.run_visit(node.type_comment, mode="func_type")

     def visit_Str(self, node: ast.Str) -> None:
         # only not PY38_PLUS
@@ -189,20 +191,6 @@ class Scanner(ast.NodeVisitor):
     def visit_Name(self, node: ast.Name) -> None:
         self.names.append(Name(lineno=node.lineno, name=node.id))

-    def iter_type_comments(self):
-        """This feature is only available for python 3.8.
-
-        PEP 526 -- Syntax for Variable Annotations
-        https://www.python.org/dev/peps/pep-0526/
-        https://docs.python.org/3.8/library/ast.html#ast.parse
-        """
-        buffer = io.StringIO(self.source)
-        for token in tokenize.generate_tokens(buffer.readline):
-            if token.type == tokenize.COMMENT:
-                comment_string = token.string.split("# type: ")
-                if comment_string != [token.string]:
-                    self.run_visit(comment_string[1], mode="func_type")
-
     @recursive
     def visit_Attribute(self, node: ast.Attribute) -> None:
         local_attr = []
@@ -222,6 +210,8 @@ class Scanner(ast.NodeVisitor):
             for item in node.value.elts:
                 if isinstance(item, ast.Str):
                     self.names.append(Name(lineno=node.lineno, name=item.s))
+        if hasattr(node, "type_comment") and node.type_comment:
+            self.run_visit(node.type_comment)

     def visit_Try(self, node: ast.Try) -> None:
         def any_import_error(items) -> bool:
@@ -246,16 +236,20 @@ class Scanner(ast.NodeVisitor):
         self.source = source
         if self.skip_file():
             return
-        if PY38_PLUS:
-            self.iter_type_comments()
         self.run_visit(self.source)
         self.import_names = [imp.name for imp in self.imports]
         self.names = list(self.get_names())
         self.unused_imports = list(self.get_unused_imports())

     def run_visit(self, source: Union[str, bytes], mode: str = "exec") -> None:
+        params = {
+            "source": source,
+            "mode": mode,
+        }
+        if PY38_PLUS:
+            params["type_comments"] = True
         try:
-            tree = ast.parse(source, mode=mode)
+            tree = ast.parse(**params)
         except SyntaxError as err:
             if self.show_error:
                 print(Color(str(err)).red)
```

**Comment by hadizakialqattan** [comment]
Your experiment is very useful, Thanks!

**Comment by hadialqattan** [comment]
Fixed: #95

**Comment by hadialqattan** [comment]
Merged: #95
